### PR TITLE
feat(gpt-oss): add FP8 reshape-cache op for SM100

### PR DESCRIPTION
**Pull request by anish-shanbhag**
## Description

This PR adds a GPT-OSS-specific SM100+ FP8 paged KV-cache update helper:

```python
flashinfer.gpt_oss_ops.reshape_and_cache_fp8(
    key,
    value,
    key_cache,
    value_cache,
    slot_mapping,
    k_scale,
    v_scale,
)
```

The op is intended for the GPT-OSS decode cache-update shape used by paged attention integrations:

- `key` / `value`: BF16 tensors with shape `[num_tokens, num_heads, 64]`
- `key_cache` / `value_cache`: FP8 E4M3 or byte-backed cache tensors with shape `[num_blocks, 16, num_heads, 64]`
- `slot_mapping`: int64 tensor with shape `[num_tokens]`; negative slots are skipped
- `k_scale` / `v_scale`: scalar FP32 scale tensors

It fuses the BF16 load, scale, E4M3 conversion, and scatter into the paged KV cache. The CUDA kernel maps one warp to one token/head row and writes two adjacent FP8 values per lane using the SM100 packed E4M3 conversion instruction.

### Motivation

GPT-OSS serving with FP8 KV cache performs this reshape/cache update on every decode step. The generic implementation is flexible, but this GPT-OSS shape is narrow and stable enough to justify a specialized path. Keeping the specialization under `flashinfer.gpt_oss_ops` avoids broadening the public generic cache API while still making the fast path available to downstream serving integrations.

### Implementation

This PR:

- Adds the SM100 CUDA implementation in `csrc/gpt_oss_reshape_cache_fp8.cu`
- Exposes a Python wrapper at `flashinfer.gpt_oss_ops.reshape_and_cache_fp8`
- Adds a JIT loader in `flashinfer.jit.gpt_oss`
- Registers the module in `flashinfer.aot` so prebuilt packages can include the op instead of requiring runtime compilation
- Adds CUDA tests covering byte-backed and `torch.float8_e4m3fn` cache tensors, negative slot skipping, and a non-contiguous value input

The wrapper performs strict validation of dtype, layout, head dimension, block size, architecture, and contiguity assumptions so unsupported cases can continue to use the existing generic path.

### Performance context

On B200/SM100, a standalone microbenchmark of the cache update path showed about **1.60x geometric-mean speedup** over the existing reshape/cache implementation across 149 representative GPT-OSS FP8 KV-cache update cases.

In a downstream vLLM GPT-OSS-120B serving integration on SM100-class GPUs with FP8 KV cache, this op produced low-single-digit end-to-end improvements at selected low-concurrency 1k input / 1k output serving points. `Tok/s/user` is latency-derived from time per output token.

Downstream integration used for these serving measurements: [anish-shanbhag/vllm@98202d867](https://github.com/anish-shanbhag/vllm/commit/98202d8679f53f3fc9eb69bdb685bbf8a27e419d).

| Parallelism | Concurrency | Total tok/s | Tok/s/GPU | Tok/s/user |
| --- | ---: | ---: | ---: | ---: |
| TP4 / 4 GPUs | 4 | 3100.2 -> 3149.8 (+1.60%) | 775.1 -> 787.5 (+1.60%) | 400.031 -> 405.004 (+1.24%) |
| TP8 / 8 GPUs | 4 | 3251.5 -> 3301.3 (+1.53%) | 406.4 -> 412.7 (+1.53%) | 421.144 -> 428.568 (+1.76%) |
| TP4 / 4 GPUs | 8 | 5534.5 -> 5604.1 (+1.26%) | 1383.6 -> 1401.0 (+1.26%) | 359.195 -> 365.221 (+1.68%) |
| TP4 / 4 GPUs | 16 | 9755.4 -> 9827.7 (+0.74%) | 2438.9 -> 2456.9 (+0.74%) | 311.999 -> 316.426 (+1.42%) |

The end-to-end benefit depends on serving shape, parallelism, and the integration-side dispatch heuristic, so the integration keeps the generic fallback path available.

## Related Issues

None.

## Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

Targeted local checks run:

```bash
python -m py_compile flashinfer/gpt_oss_ops/kv_cache.py flashinfer/gpt_oss_ops/__init__.py flashinfer/jit/gpt_oss.py flashinfer/jit/__init__.py flashinfer/aot.py tests/gpt_oss_ops/test_reshape_cache_fp8.py
ruff check flashinfer/gpt_oss_ops/kv_cache.py flashinfer/gpt_oss_ops/__init__.py flashinfer/jit/gpt_oss.py flashinfer/jit/__init__.py flashinfer/aot.py tests/gpt_oss_ops/test_reshape_cache_fp8.py
git diff --check
```

### Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Added `tests/gpt_oss_ops/test_reshape_cache_fp8.py`. The CUDA tests skip when CUDA, SM100+, or `torch.float8_e4m3fn` support is unavailable.

## Reviewer Notes

This API is intentionally placed under `gpt_oss_ops` because the kernel specializes for GPT-OSS cache layout assumptions rather than a general paged-cache update contract. Downstream integrations should continue to guard on dtype, shape, architecture, and serving-shape heuristics before dispatching to this fast path.

## Summary by CodeRabbit

* **New Features**
  * Added `reshape_and_cache_fp8` to convert GPT-OSS BF16 key/value tensors into FP8 (E4M3) paged KV caches using `slot_mapping`, with support for per-tensor scaling.
  * Automatically enables the optimized implementation and JIT module generation on SM100+ GPUs, and exposes the operator through the GPT-OSS ops API.
* **Tests**
  * Added SM100+-gated CUDA tests validating correctness across head counts, cache dtypes, and scale shapes, plus `fi_trace` coverage.
* **Refactor / Chores**
  * Added a GPT-OSS FP8 reshape-and-cache trace template and updated test imports to ensure trace templates are registered consistently.

### CODE REVIEW

File: csrc/gpt_oss_reshape_cache_fp8.cu [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+  key_float.x *= kinv;
+  key_float.y *= kinv;
+  value_float.x *= vinv;
+  value_float.y *= vinv;
+
+  const uint16_t ko = cvt_e4m3x2(key_float.y, key_float.x);
+  const uint16_t vo = cvt_e4m3x2(value_float.y, value_float.x);
+
+  const int64_t block = slot / block_size;
+  const int64_t offset = slot - block * block_size;
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply csrc/gpt_oss_reshape_cache_fp8.cu]: Fixed in 24df2bed: the host path still validates the page size as 16, and the kernel now uses `kPageSize` / `kPageSizeLog2` with shift + mask for the block and offset calculation instead of runtime divide/mod.

File: csrc/gpt_oss_reshape_cache_fp8.cu [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  TVM_FFI_ICHECK_EQ(value.dtype(), dl_bfloat16) << "value must be bfloat16";
+  TVM_FFI_ICHECK(key_cache.dtype() == dl_uint8 || key_cache.dtype() == dl_float8_e4m3fn)
+      << "key_cache must be uint8 or float8_e4m3fn";
+  TVM_FFI_ICHECK(value_cache.dtype() == dl_uint8 || value_cache.dtype() == dl_float8_e4m3fn)
+      << "value_cache must be uint8 or float8_e4m3fn";
+  TVM_FFI_ICHECK_EQ(slot_mapping.dtype(), dl_int64) << "slot_mapping must be int64";
+  TVM_FFI_ICHECK_EQ(k_scale.dtype(), dl_float32) << "k_scale must be float32";
+  TVM_FFI_ICHECK_EQ(v_scale.dtype(), dl_float32) << "v_scale must be float32";
+  TVM_FFI_ICHECK_EQ(k_scale.size(0), 1) << "k_scale must have one element";
+  TVM_FFI_ICHECK_EQ(v_scale.size(0), 1) << "v_scale must have one element";
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply csrc/gpt_oss_reshape_cache_fp8.cu]: Fixed in 24df2bed: the wrapper now accepts either a rank-0 scalar scale tensor or a rank-1 single-element scale tensor, and only calls `size(0)` after checking `ndim()`.

File: csrc/gpt_oss_reshape_cache_fp8.cu [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+  TVM_FFI_ICHECK_EQ(key_cache.size(2), num_heads) << "key_cache head count mismatch";
+  TVM_FFI_ICHECK_EQ(value_cache.size(2), num_heads) << "value_cache head count mismatch";
+  TVM_FFI_ICHECK_EQ(key_cache.size(3), 64) << "key_cache head_dim must be 64";
+  TVM_FFI_ICHECK_EQ(value_cache.size(3), 64) << "value_cache head_dim must be 64";
+  TVM_FFI_ICHECK_GE(key.stride(1), 64) << "key head stride must be at least 64";
+  TVM_FFI_ICHECK_GE(value.stride(1), 64) << "value head stride must be at least 64";
+  TVM_FFI_ICHECK_EQ(key_cache.stride(1), 64) << "key_cache token stride must be 64";
+  TVM_FFI_ICHECK_EQ(value_cache.stride(1), 64) << "value_cache token stride must be 64";
+  TVM_FFI_ICHECK_GE(key_cache.stride(2), 64) << "key_cache head stride must be at least 64";
+  TVM_FFI_ICHECK_GE(value_cache.stride(2), 64) << "value_cache head stride must be at least 64";
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply csrc/gpt_oss_reshape_cache_fp8.cu:216]: Addressed in 24df2bed: added host-side pointer alignment checks for the `bfloat162` loads and `uint16_t` stores, plus even-stride checks for the token/head/cache strides that feed those vectorized accesses.

File: flashinfer/aot.py [hunk 4]
```diff
@@ -585,6 +586,8 @@ def gen_all_modules(
             gen_sampling_module(),
             gen_topk_module(),
         ]
+        if has_sm100 or has_sm103 or has_sm110 or has_sm120 or has_sm121:
+            jit_specs.append(gen_gpt_oss_reshape_cache_fp8_module())
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply flashinfer/aot.py]: Done in 24df2bed: added `has_sm100f` and `has_sm120f` explicitly to the AOT condition.

File: tests/gpt_oss_ops/test_reshape_cache_fp8.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        (num_tokens, 10, num_heads, head_dim),
+        device="cuda:0",
+        dtype=torch.bfloat16,
+    )
+    value = value_storage[:, 0]
+    slot_mapping = torch.tensor(
+        [0, 1, 2, 15, -1, 16, 31, 33], device="cuda:0", dtype=torch.int64
+    )
+    k_scale = torch.tensor([1.0], device="cuda:0", dtype=torch.float32)
+    v_scale = torch.tensor([1.0], device="cuda:0", dtype=torch.float32)
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply tests/gpt_oss_ops/test_reshape_cache_fp8.py]: Done in 24df2bed: the CUDA test now uses non-trivial `1.5` / `0.8` scales and covers both rank-0 scalar and rank-1 single-element scale tensors.

File: csrc/gpt_oss_reshape_cache_fp8.cu [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  const int64_t offset = slot - block * block_size;
+  // Cache tensors are laid out as [num_blocks, 16, num_heads, 64]. Strides are
+  // byte offsets because the cache may be passed either as uint8 storage or as
+  // torch.float8_e4m3fn.
+  const int64_t key_dst =
+      block * key_block_stride + offset * key_page_stride + head * key_head_cache_stride;
+  const int64_t value_dst =
+      block * value_block_stride + offset * value_page_stride + head * value_head_cache_stride;
+  reinterpret_cast<uint16_t*>(key_cache + key_dst)[lane] = ko;
+  reinterpret_cast<uint16_t*>(value_cache + value_dst)[lane] = vo;
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply csrc/gpt_oss_reshape_cache_fp8.cu]: Fixed in 24df2bed: the host passes `max_slots = num_blocks * block_size`, and the kernel now returns before loading/scattering when `slot < 0 || slot >= max_slots`. The test also includes an out-of-range slot to cover this path.

File: csrc/gpt_oss_reshape_cache_fp8.cu [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+  // enough independent (token, head) work to fill them.
+  if (total_items <= 1) {
+    LAUNCH(1);
+  } else if (total_items <= 2) {
+    LAUNCH(2);
+  } else if (total_items <= 4) {
+    LAUNCH(4);
+  } else {
+    LAUNCH(8);
+  }
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply csrc/gpt_oss_reshape_cache_fp8.cu]: Addressed in 24df2bed: replaced the local launch macro plus explicit if-ladder with a `DISPATCH_WARPS_PER_CTA` dispatch macro for the 1/2/4/8 warp launch variants.

File: flashinfer/gpt_oss_ops/kv_cache.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+@flashinfer_api
+def reshape_and_cache_fp8(
+    key: torch.Tensor,
+    value: torch.Tensor,
+    key_cache: torch.Tensor,
+    value_cache: torch.Tensor,
+    slot_mapping: torch.Tensor,
+    k_scale: torch.Tensor,
+    v_scale: torch.Tensor,
+) -> None:
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply flashinfer/gpt_oss_ops/kv_cache.py]: Done in 24df2bed: added a GPT-OSS TraceTemplate, wired the public API as `@flashinfer_api(trace=gpt_oss_reshape_cache_fp8_trace)`, and added trace smoke/registry coverage.

File: flashinfer/jit/gpt_oss.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+from . import env as jit_env
+
+
+def gen_gpt_oss_reshape_cache_fp8_module() -> JitSpec:
+    return gen_jit_spec(
+        "gpt_oss_reshape_cache_fp8",
+        [
+            jit_env.FLASHINFER_CSRC_DIR / "gpt_oss_reshape_cache_fp8.cu",
+        ],
+    )
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply flashinfer/jit/gpt_oss.py:14]: Done in 24df2bed: imported `functools` and decorated `gen_gpt_oss_reshape_cache_fp8_module()` with `@functools.cache`.

File: tests/gpt_oss_ops/test_reshape_cache_fp8.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+from flashinfer.gpt_oss_ops import reshape_and_cache_fp8
+
+
+def _require_sm100():
+    if not torch.cuda.is_available():
+        pytest.skip("CUDA is required")
+    major, _ = torch.cuda.get_device_capability(0)
+    if major < 10:
+        pytest.skip("reshape_and_cache_fp8 requires SM100 or newer")
+
```
[the remark this answers is not in the document]
Comment (anish-shanbhag) [reply tests/gpt_oss_ops/test_reshape_cache_fp8.py:22]: Done in 24df2bed: the test gate now uses the FlashInfer architecture helpers for SM100/SM110/SM12x support instead of raw capability-number checks.

**Review by nv-yunzheq — APPROVED** [review]
